internal/resources/aws/sns_topic.go

```diff
@@ -1,62 +1,230 @@
+	return r.notificationsCostComponent(
+		"HTTP/HTTPS notifications (over 100k)",
+		"100k notifications",
+		100000,
+		"DeliveryAttempts-HTTP",
+		100000,
+		subscriptions,
+		requests,
+	)
+}
+
+func (r *SNSTopic) EmailNotificationsCostComponent(subscriptions, requests *int64) *schema.CostComponent {
+	return r.notificationsCostComponent(
+		"Email/Email-JSON notifications (over 1k)",
+		"100k notifications",
+		100000,
+		"DeliveryAttempts-SMTP",
+		1000,
+		subscriptions,
+		requests,
+	)
+}
+
+func (r *SNSTopic) KinesisNotificationsCostComponent(subscriptions, requests *int64) *schema.CostComponent {
+	return r.notificationsCostComponent(
+		"Kinesis Firehose notifications",
+		"1M notifications",
+		1000000,
+		"DeliveryAttempts-FIREHOSE",
+		0,
+		subscriptions,
+		requests,
+	)
+}
+
+func (r *SNSTopic) MobilePushNotificationsCostComponent(subscriptions, requests *int64) *schema.CostComponent {
+	return r.notificationsCostComponent(
+		"Mobile Push notifications",
+		"1M notifications",
+		1000000,
+		"DeliveryAttempts-APNS",
+		0,
+		subscriptions,
+		requests,
+	)
+}
+
+func (r *SNSTopic) MacOSNotificationsCostComponent(subscriptions, requests *int64) *schema.CostComponent {
+	return r.notificationsCostComponent(
+		"MacOS notifications",
+		"1M notifications",
+		1000000,
+		"DeliveryAttempts-MACOS",
+		0,
+		subscriptions,
+		requests,
+	)
+}
+
+func (r *SNSTopic) SMSNotificationsCostComponent(subscriptions, requests *int64, customPrice *float64) *schema.CostComponent {
+	c := r.notificationsCostComponent(
+		"SMS notifications (over 100)",
+		"100 notifications",
+		100,
+		"DeliveryAttempts-SMS",
+		100,
+		subscriptions,
+		requests,
+	)
+
+	if customPrice != nil {
+		c.SetCustomPrice(decimalPtr(decimal.NewFromFloat(*customPrice)))
+	}
+
+	return c
 }
 
 func (r *SNSTopic) PopulateUsage(u *schema.UsageData) {
 	resources.PopulateArgsWithUsage(r, u)
 }
 
 func (r *SNSTopic) BuildResource() *schema.Resource {
-	var requests *decimal.Decimal
+	var requests *int64
 
-	requestSize := decimal.NewFromInt(64)
+	requestSize := 64.0
 	if r.RequestSizeKB != nil {
-		requestSize = decimal.NewFromFloat(*r.RequestSizeKB)
+		requestSize = *r.RequestSizeKB
 	}
 
 	if r.MonthlyRequests != nil {
-		requests = decimalPtr(r.calculateRequests(requestSize, decimal.NewFromInt(*r.MonthlyRequests)))
+		requests = r.calculateRequests(requestSize, *r.MonthlyRequests)
+	}
+
+	components := []*schema.CostComponent{
+		r.APIRequestsCostComponent(requests),
+		r.HTTPNotificationsCostComponent(r.HTTPSubscriptions, requests),
+		r.EmailNotificationsCostComponent(r.EmailSubscriptions, requests),
+		r.KinesisNotificationsCostComponent(r.KinesisSubscriptions, requests),
+		r.MobilePushNotificationsCostComponent(r.MobilePushSubscriptions, requests),
+		r.MacOSNotificationsCostComponent(r.MacOSSubscriptions, requests),
+		r.SMSNotificationsCostComponent(r.SMSSubscriptions, requests, r.SMSNotificationPrice),
```

Comment:
I'm not sure why we are passing in struct properties into struct methods. Could we not just do:

```suggestion
		r.HTTPNotificationsCostComponent(requests),
		r.EmailNotificationsCostComponent(requests),
		r.KinesisNotificationsCostComponent(requests),
		r.MobilePushNotificationsCostComponent(requests),
		r.MacOSNotificationsCostComponent(requests),
		r.SMSNotificationsCostComponent(requests),
```

and then the struct methods would fetch the relevant property, e.g:

```go

func (r *SNSTopic) EmailNotificationsCostComponent(requests *int64) *schema.CostComponent {
	return r.notificationsCostComponent(
		"Email/Email-JSON notifications (over 1k)",
		"100k notifications",
		100000,
		"DeliveryAttempts-SMTP",
		1000,
		r.EmailSubscriptions,
		requests,
	)
}
```

Reply:
I considered that, but I prefer the clarity of passing them in explicitly.